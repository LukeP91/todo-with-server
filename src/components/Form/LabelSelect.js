import _ from 'lodash'
import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import FormControl from '@material-ui/core/FormControl'
import Select from '@material-ui/core/Select'
import Chip from '@material-ui/core/Chip'

const styles = theme => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  formControl: {
    margin: theme.spacing.unit,
    minWidth: 120,
    maxWidth: 300,
  },
  chips: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  chip: {
    margin: theme.spacing.unit / 4,
  },
})

const ITEM_HEIGHT = 48
const ITEM_PADDING_TOP = 8
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 250,
    },
  },
}

class MultipleSelect extends React.Component {
  selected = () => {
    const {
      labels,
      classes,
      form: { values },
    } = this.props
    return labels
      .filter(label => _.includes(values.labelsIds, label.id))
      .map(label => <Chip className={classes.chip} key={label.id} label={label.value} />)
  }

  render() {
    const {
      field: { name, ...field },
      theme,
      classes,
      labels,
      form: { values },
    } = this.props

    return (
      <div className={classes.root}>
        <FormControl className={classes.formControl} id="labels-list">
          <InputLabel htmlFor="select-multiple-chip">Labels</InputLabel>
          <Select
            {...field}
            id={name}
            MenuProps={MenuProps}
            multiple
            name={name}
            renderValue={() => <div className={classes.chips}>{this.selected()}</div>}
          >
            {labels.map(label => (
              <MenuItem
                key={label.id}
                style={{
                  fontWeight:
                    values.labelsIds.indexOf(label.id) === -1
                      ? theme.typography.fontWeightRegular
                      : theme.typography.fontWeightMedium,
                }}
                value={label.id}
              >
                {label.value}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      </div>
    )
  }
}

MultipleSelect.propTypes = {
  classes: PropTypes.object.isRequired,
  field: PropTypes.object,
  labels: PropTypes.array.isRequired,
  theme: PropTypes.object,
}

export default withStyles(styles, { withTheme: true })(MultipleSelect)
