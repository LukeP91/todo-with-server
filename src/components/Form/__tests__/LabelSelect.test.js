import React from 'react'
import renderer from 'react-test-renderer'
import Chip from '@material-ui/core/Chip'

import LabelSelect from '../LabelSelect'

describe('LabelSelect', () => {
  it('renders correctly', () => {
    const props = {
      labels: [{ id: 1, value: 'Label1' }, { id: 2, value: 'Label2' }],
      form: { values: { labelsIds: [1] } },
      field: { name: 'test', value: [1] },
    }
    const tree = renderer.create(<LabelSelect {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('displayes selected labels as chips', () => {
    const props = {
      labels: [{ id: 1, value: 'Label1' }, { id: 2, value: 'Label2' }, { id: 3, value: 'Label2' }],
      form: { values: { labelsIds: [1, 3] } },
      field: { name: 'test', value: [1, 3] },
    }

    const tree = renderer.create(<LabelSelect {...props} />)
    const elements = tree.root.findAllByType(Chip).map(chip => chip.props.label)

    expect(elements).toEqual(['Label1', 'Label2'])
  })
})
