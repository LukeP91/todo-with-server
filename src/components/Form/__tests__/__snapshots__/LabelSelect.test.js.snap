// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelSelect renders correctly 1`] = `
<div
  className="MultipleSelect-root-1"
>
  <div
    className="MuiFormControl-root-5 MultipleSelect-formControl-2"
  >
    <label
      className="MuiFormLabel-root-14 MuiFormLabel-filled-18 MuiInputLabel-root-9 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
      data-shrink={true}
      htmlFor="select-multiple-chip"
    >
      Labels
    </label>
    <div
      className="MuiInput-root-26 MuiInput-formControl-27 MuiInput-underline-30"
    >
      <div
        className="MuiSelect-root-21"
      >
        <div
          aria-haspopup="true"
          aria-owns={null}
          aria-pressed="false"
          className="MuiSelect-select-22 MuiSelect-selectMenu-23 MuiInput-input-34"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MultipleSelect-chips-3"
          >
            <div
              className="MuiChip-root-39 MultipleSelect-chip-4"
              onKeyDown={[Function]}
              role="button"
              tabIndex={-1}
            >
              <span
                className="MuiChip-label-52"
              >
                Label1
              </span>
            </div>
          </div>
        </div>
        <input
          id="test"
          name="test"
          type="hidden"
          value="1"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-56 MuiSelect-icon-25"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
